### ISSUE TITLE
Test suite occasionally fails, with a segfault at 0x0

### ISSUE THREAD

**Issue by 64kramsystem** [tenderlove/tenderjit#110]
I've noticed, while running the test suite for a large amount of sequential runs, that TJ sometimes segfaults.

The failure is very nondeterministic. Sometimes it takes a few runs, sometimes it doesn't happen in a hundred runs. It does not depend on the test seed.

Since my last two PRs were quite sensitive, I've checked if the issue was present before they were merged, and I can confirm that the issue was already present.

Below there are some sample failures; I think the only pointers (haha) they give is that, since the segfault address is 0, this should be either a null pointer, or I think more likely, a misaligned stack.

A very long run on Mac may confirm if this is Linux-only, or cross-platform. But unfortunately, since even 100 runs don't guarantee a failure, it may hard to reproduce.

Sample 1:

```
......S...........SS......SSS.S....S...S.SS....S.S..S../home/saverio/code/fisk-dev/lib/fisk.rb:847: [BUG] Segmentation fault at 0x0000000000000000
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0040 p:---- s:0223 e:000222 CFUNC  :zip
c:0039 p:0023 s:0218 e:000217 BLOCK  /home/saverio/code/fisk-dev/lib/fisk.rb:847 [FINISH]
c:0038 p:---- s:0214 e:000213 IFUNC 
c:0037 p:---- s:0211 e:000210 CFUNC  :each
c:0036 p:---- s:0208 e:000207 CFUNC  :find_all
c:0035 p:0007 s:0204 e:000203 METHOD /home/saverio/code/fisk-dev/lib/fisk.rb:845
c:0034 p:0022 s:0192 e:000191 METHOD /home/saverio/code/fisk-dev/lib/fisk/instructions.rb:1400
c:0033 p:0014 s:0187 e:000186 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/runtime.rb:198
c:0032 p:0049 s:0181 e:000180 BLOCK  /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:2312
c:0031 p:0038 s:0177 e:000176 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:3001
c:0030 p:0011 s:0171 e:000170 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:2307
c:0029 p:0360 s:0166 e:000165 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:130
c:0028 p:0137 s:0154 e:000153 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:84
c:0027 p:0148 s:0149 e:000148 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit.rb:558
c:0026 p:0009 s:0140 e:000139 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:557 [FINISH]
Segmentation fault
```

Sample 2:

```
.S..S.............S.SS.S.S....SSS...SS...................SS.........SS......./home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:45: [BUG] Segmentation fault at 0x0000000000000000
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0029 p:0294 s:0163 e:000162 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:45 [FINISH]
c:0028 p:---- s:0155 e:000154 CFUNC  :new
c:0027 p:0125 s:0149 e:000148 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit.rb:552
c:0026 p:0009 s:0140 e:000139 METHOD /home/saverio/code/tenderjit-dev/lib/tenderjit/iseq_compiler.rb:557 [FINISH]
Segmentation fault
```

**Comment by tenderlove** [comment]
I bet this is to do with the `expandarray` patches I added (that's my first hunch anyway)

**Comment by tenderlove** [comment]
Good news!  I can reliably reproduce Linux like this: `bundle exec rake test TEST_OPTS='--seed=58785'`.

I'm not sure what the issue is though. 😅

**Comment by 64kramsystem** [comment]
> Good news! I can reliably reproduce Linux like this: `bundle exec rake test TEST_OPTS='--seed=58785'`.

Thanks for trying :blush: 

Having said that... rats, I can't reproduce it :cursing_face: :exploding_head: 

> I'm not sure what the issue is though.

It's possible to potentially restrict the cause via LLDB. If during a test run, the instruction causing the segfault is something like a `MOVASD`, then we it's a stack misalignment problem - this instruction (class) segfaults when executing on a misaligned stack (`mov` `a`ligned `s`calar `d`ouble precision floating point¹). I'll see if I confirm this.

It's hard for me to bisect, though. I should probably run 500/1000 cycles per commit :joy: Not impossible, I will probably do it at some point, but it's certainly not a standard test session :smile: 

¹=I've just found that there's also a `MOVASS` instruction :joy:

**Comment by 64kramsystem** [comment]
So, just brainstorming a bit :smile: 

With this diff:

```diff
diff --git i/Rakefile w/Rakefile
index 09d9d86..2c7e2d9 100644
--- i/Rakefile
+++ w/Rakefile
@@ -88,15 +88,15 @@ end
 #   in this case, only one file must be found.
 #
 task :debug, [:test_suite_file] => test_files + [:compile] do |_, args|
-  test_suite_file =
-    case args.test_suite_file
-    when %r{/}
-      args.test_suite_file
-    else
-      files = FileList["test/**/#{args.test_suite_file}_test.rb"]
-      raise "Only one file must be found (found: #{files}" if files.size != 1
-      files.first
-    end
+  # test_suite_file =
+  #   case args.test_suite_file
+  #   when %r{/}
+  #     args.test_suite_file
+  #   else
+  #     files = FileList["test/**/#{args.test_suite_file}_test.rb"]
+  #     raise "Only one file must be found (found: #{files}" if files.size != 1
+  #     files.first
+  #   end
 
   library_dirs = %w[lib test] + DEBUG_LIBRARIES.map do |lib|
     "#{Gem::Specification.find_by_name(lib).gem_dir}/lib"
@@ -118,9 +118,11 @@ task :debug, [:test_suite_file] => test_files + [:compile] do |_, args|
   # -v: verbose mode
   # -I: load paths
   #
-  Bundler.with_unbundled_env do
-    command = "lldb -o run ruby -- -d -v -I #{library_dirs.join(":")} #{test_suite_file}"
-    system command
+  100.times do
+    Bundler.with_unbundled_env do
+      command = "lldb -o 'settings set target.process.stop-on-exec false' -o run -o q ruby -- -d -v -I #{library_dirs.join(":")} $(which bundle) exec rake test"
+      system command
+    end
   end
 end
 ```

The command `rake debug` will run the test suite 100 times with lldb attached.

There's a missing piece though - since child processes are being executed, lldb is actually not attached, and when the test suite segfaults, lldb won't show the code at RIP.

If you know on the spot how to deal with this, it'd save me some head bashing :laughing:, as I've never used lldb before TJ :smile:

If I manage to implement this functionality, I could actually remove the `rake debug` restriction of having to specify a test suite, making it functionally equivalent to `rake test`. And also, probably make the `-o q` smarter, and exit only if the process exited cleanly. All of this would make debugging TJ easier.

**Comment by tenderlove** [comment]
@64kramsystem I'll take a look at this today.  I've been meaning to make a debug task because getting TJ running under a debugger isn't really the easiest 😅

**Comment by tenderlove** [comment]
@64kramsystem here's what I've learned so far wrt lldb (I'm still investigating the crash).  If we specify a [`RUBY` environment variable](https://github.com/ruby/rake/blob/master/lib/rake/file_utils.rb#L9-L13), Rake will [use that as the Ruby](https://github.com/ruby/rake/blob/master/lib/rake/file_utils.rb#L102-L104).  So it's possible to do this:

```
$ RUBY='lldb ruby --' rake test
```

And that will load the full suite under lldb.  I also tried doing this:

```
aaron@whiteclaw ~/g/tenderjit (main)> git diff
diff --git a/Rakefile b/Rakefile
index 09d9d86..48bb88c 100644
--- a/Rakefile
+++ b/Rakefile
@@ -79,6 +79,10 @@ task :test, [:test_suite_prefix, :test_bare_name] => test_files + [:compile] do
   end
 end
 
+task :lldb do
+  FileUtils.const_set(:RUBY, "lldb #{FileUtils::RUBY} --")
+end
+
 # Launch an LLDB debug session.
 #
 # Don't forget to put an `rt.break`!
```

Then you can do `rake lldb test` and it will do the same thing (though I feel weird about doing the const set).  We could probably set the env var then shell out again to `rake test`, but honestly I don't feel *that* bad about doing a `const_set` 😆

**Comment by tenderlove** [comment]
I think I found it! [Here](https://github.com/tenderlove/tenderjit/pull/111).  It looks like we were calling a cfunc and manually preserving a register even though calling the cfunc preserved it for us.  Does that PR seem to fix things for you?

The way I found this was I got a backtrace under lldb:

```
(lldb) bt
* thread #1, name = 'ruby', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x00005555555b8730 ruby`rbimpl_size_mul_overflow(x=1, y=40) at memory.h:568:41
    frame #1: 0x00005555555a67c1 ruby`size_mul_or_raise(x=1, y=40, exc=0x00007ffff7513228) at gc.c:183:45
    frame #2: 0x00005555555b1417 ruby`xmalloc2_size(count=1, elsize=40) at gc.c:11451:12
    frame #3: 0x00005555555b1450 ruby`ruby_xcalloc_body(n=1, size=40) at gc.c:11703:43
    frame #4: 0x00005555555a77dd ruby`ruby_xcalloc(n=1, size=40) at gc.c:13670:12
    frame #5: 0x00005555555a789a ruby`rb_data_typed_object_zalloc(klass=0x00007ffff7528380, size=40, type=0x0000555555a99028) at gc.c:2768:21
    frame #6: 0x00005555556b8d81 ruby`rb_proc_alloc(klass=0x00007ffff7528380) at proc.c:167:12
    frame #7: 0x0000555555808283 ruby`vm_proc_create_from_captured(klass=0x00007ffff7528380, captured=0x00007ffff7664a68, block_type=block_type_iseq, is_from_method='\0', is_lambda='\0') at vm.c:938:21
    frame #8: 0x0000555555808064 ruby`rb_vm_make_proc_lambda(ec=0x0000555555ab7720, captured=0x00007ffff7664a68, klass=0x00007ffff7528380, is_lambda='\0') at vm.c:1216:15
    frame #9: 0x00005555557d52f7 ruby`rb_vm_make_proc(ec=0x0000555555ab7720, captured=0x00007ffff7664a68, klass=0x00007ffff7528380) at vm_core.h:1737:12
    frame #10: 0x00005555557d51b0 ruby`rb_vm_bh_to_procval(ec=0x0000555555ab7720, block_handler=0x00007ffff7664a69) at vm_insnhelper.c:486:13
    frame #11: 0x00007ffff255e14d
    frame #12: 0x000055555580acd0 ruby`rb_vm_exec(ec=0x0000555555ab7720, mjit_enable_p=true) at vm.c:2208:41
    frame #13: 0x00005555558320f6 ruby`invoke_block(ec=0x0000555555ab7720, iseq=0x00007ffff3b5e988, self=0x00007ffff3e2d038, captured=0x00007ffff7664a28, cref=0x0000000000000000, type=0x0000000022220001, opt_pc=0) at vm.c:1314:12
```

Noticed that we have a JIT frame calling in to `rb_vm_bh_to_procval`.  I found where we did that, set a breakpoint, then checked the alignment before calling in to the function.  Also I saw that we were pushing `r9` twice (which we probably didn't want to do)

**Comment by 64kramsystem** [comment]
> @64kramsystem here's what I've learned so far wrt lldb (I'm still investigating the crash). If we specify a [`RUBY` environment variable](https://github.com/ruby/rake/blob/master/lib/rake/file_utils.rb#L9-L13), Rake will [use that as the Ruby](https://github.com/ruby/rake/blob/master/lib/rake/file_utils.rb#L102-L104). So it's possible to do this:

Bwaahahaha, that's clever :laughing:. I agree that the const set is preferrable/acceptable due to its simplicity.

**Comment by 64kramsystem** [comment]
> I think I found it! [Here](https://github.com/tenderlove/tenderjit/pull/111). It looks like we were calling a cfunc and manually preserving a register even though calling the cfunc preserved it for us. Does that PR seem to fix things for you?
> 
> The way I found this was I got a backtrace under lldb:
> 
> ```
> (lldb) bt
> * thread #1, name = 'ruby', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
>   * frame #0: 0x00005555555b8730 ruby`rbimpl_size_mul_overflow(x=1, y=40) at memory.h:568:41
>     frame #1: 0x00005555555a67c1 ruby`size_mul_or_raise(x=1, y=40, exc=0x00007ffff7513228) at gc.c:183:45
>     frame #2: 0x00005555555b1417 ruby`xmalloc2_size(count=1, elsize=40) at gc.c:11451:12
>     frame #3: 0x00005555555b1450 ruby`ruby_xcalloc_body(n=1, size=40) at gc.c:11703:43
>     frame #4: 0x00005555555a77dd ruby`ruby_xcalloc(n=1, size=40) at gc.c:13670:12
>     frame #5: 0x00005555555a789a ruby`rb_data_typed_object_zalloc(klass=0x00007ffff7528380, size=40, type=0x0000555555a99028) at gc.c:2768:21
>     frame #6: 0x00005555556b8d81 ruby`rb_proc_alloc(klass=0x00007ffff7528380) at proc.c:167:12
>     frame #7: 0x0000555555808283 ruby`vm_proc_create_from_captured(klass=0x00007ffff7528380, captured=0x00007ffff7664a68, block_type=block_type_iseq, is_from_method='\0', is_lambda='\0') at vm.c:938:21
>     frame #8: 0x0000555555808064 ruby`rb_vm_make_proc_lambda(ec=0x0000555555ab7720, captured=0x00007ffff7664a68, klass=0x00007ffff7528380, is_lambda='\0') at vm.c:1216:15
>     frame #9: 0x00005555557d52f7 ruby`rb_vm_make_proc(ec=0x0000555555ab7720, captured=0x00007ffff7664a68, klass=0x00007ffff7528380) at vm_core.h:1737:12
>     frame #10: 0x00005555557d51b0 ruby`rb_vm_bh_to_procval(ec=0x0000555555ab7720, block_handler=0x00007ffff7664a69) at vm_insnhelper.c:486:13
>     frame #11: 0x00007ffff255e14d
>     frame #12: 0x000055555580acd0 ruby`rb_vm_exec(ec=0x0000555555ab7720, mjit_enable_p=true) at vm.c:2208:41
>     frame #13: 0x00005555558320f6 ruby`invoke_block(ec=0x0000555555ab7720, iseq=0x00007ffff3b5e988, self=0x00007ffff3e2d038, captured=0x00007ffff7664a28, cref=0x0000000000000000, type=0x0000000022220001, opt_pc=0) at vm.c:1314:12
> ```
> 
> Noticed that we have a JIT frame calling in to `rb_vm_bh_to_procval`. I found where we did that, set a breakpoint, then checked the alignment before calling in to the function. Also I saw that we were pushing `r9` twice (which we probably didn't want to do)

So!

That's definitely a valid fix (I'm adding some considerations in the PR). It seems that it's not the cause of the problem, though :exploding_head: :sob: 

Running from the PR, I still experience occasional crashes. Thanks to the new lldb/ruby superpowers :grin:, I get more clarity:

```
* thread #1, name = 'ruby', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x00007ffff7e7b785 libruby.so.3.0`rb_execution_context_mark(ec=0x000055555555d810) at vm.c:2858:20
libruby.so.3.0`rb_execution_context_mark:
->  0x7ffff7e7b785 <+341>: test   byte ptr [rdx], 0x4
    0x7ffff7e7b788 <+344>: jne    0x7ffff7e7b730            ; <+256> at vm.c:2859:21
    0x7ffff7e7b78a <+346>: test   al, 0x4
    0x7ffff7e7b78c <+348>: je     0x7ffff7e7b748            ; <+280> at vm.c:2868:10

thread #1, name = 'ruby', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x00007ffff7e7b785 libruby.so.3.0`rb_execution_context_mark(ec=0x000055555555d810) at vm.c:2858:20
    frame #1: 0x00007ffff7e7b8cd libruby.so.3.0`thread_mark(ptr=0x000055555555cfb0) at vm.c:2940:5
    frame #2: 0x00007ffff7cca81f libruby.so.3.0`gc_mark_children(objspace=0x000055555555d150, obj=0x00005555555caff8) at gc.c:6363:19
```

(I hope that this is correct; in order to be able to test in human times (:joy:), I'm running 100 `rake test`s in parallel (:joy: :joy:), so the output is scrambled).

It seems that there is a garbage collection related issue somewhere; the nature of it actually explains why the error is only sporadic (I guess that, say, generating a bazillion objects, and/or invoking the GC, may help triggering it).

Is it possible that there is a problem along the lines of what was fixed by #112?

**Comment by tenderlove** [comment]
> Is it possible that there is a problem along the lines of what was fixed by #112?

It's possible.  But looking at the stack trace it seems like maybe frame pushing is broken.  We might be sticking the wrong value in the EP, then GC happens to run when there's a frame that's been pushed by the JIT.  Also it looks like a problem with frames pushed when entering a block.

I'll try again to reproduce this.  Apparently the crash I was seeing is not this crash 😅

**Comment by tenderlove** [comment]
I'm going to be pretty busy for the next week with holiday stuff, but I've been thinking about this problem and I wonder if it's that we're pushing a block handler on the stack but not executing the write barrier?  Just a guess and a thought crossing my mind (I just wanted to write it down so when I come back to this I'll remember 😅)

**Comment by tenderlove** [comment]
I enabled VM debugging and this is definitely a GC problem.  I saw a crash like this:

```
TenderJIT::PutnilTest#test_putnil = 0.04 s = .                                                                                                                                               
TenderJIT::InternTest#test_new_symbol = 0.06 s = .                                                                                                                                           
TenderJIT::InternTest#test_existing_symbol = 0.17 s = .                                                                                                                                      
TenderJIT::IterFib#test_fib_iter = Assertion Failed: vm.inc:829:vm_exec_core:!RB_TYPE_P(TOPN(0), T_NONE)                                                                                     
ruby 3.2.0dev (2021-12-29T11:53:43Z master 002fa28599) [x86_64-linux]                                                                                                                        
                                                                                                                                                                                             
-- Control frame information -----------------------------------------------                                                                                                                 
c:0032 p:0038 s:0199 e:000196 BLOCK  /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1743                                                                                           
c:0031 p:0015 s:0193 e:000192 METHOD /home/aaron/.gem/ruby/3.0.4/bundler/gems/fisk-13a2129aea03/lib/fisk.rb:649                                                                              
c:0030 p:0065 s:0187 e:000186 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1741                                                                                           
c:0029 p:0310 s:0173 e:000172 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:130                                                                                            
c:0028 p:0025 s:0161 e:000160 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1023 [FINISH]                                                                                  
c:0027 p:0000 s:0153 e:000152 METHOD /home/aaron/git/tenderjit/test/tenderjit_test.rb:12                                                                                                     
c:0026 p:0005 s:0145 e:000144 METHOD /home/aaron/git/tenderjit/test/tend
```

I built Ruby like this:

```
begin
  set -lx debugflags '-g'
  set -lx optflags '-O0'
  set -lx cppflags '-DRUBY_DEBUG -DVM_CHECK_MODE'
  set -lx CC 'clang'
  ./configure --prefix=$HOME/.rubies/ruby-trunk --disable-install-doc
end
```

It enables VM assertions, and one of the assertions is to make sure that we never push T_NONE or T_MOVED on the stack.  We're pushing a T_NONE on the stack so clearly something is getting collected that shouldn't be.  I'm guessing we're missing a write barrier, as you suspected @64kramsystem.

**Comment by tenderlove** [comment]
Yep, we're definitely missing a write barrier and it's my fault. The `getblockparam` instruction calls `vm_env_write` and that function can possibly trigger a write barrier [here](https://github.com/ruby/ruby/blob/002fa2859962f22de8afdbeece04966ea57b7da9/vm_insnhelper.c#L465-L474).  We're [not checking that flag in our version](https://github.com/tenderlove/tenderjit/blob/d3f640e6951db41283ce76dbaad25ca7a839daf5/lib/tenderjit/iseq_compiler.rb#L2636-L2641) so we could possibly write to the EP without triggering the write barrier.

I'll work on a fix!

**Comment by tenderlove** [comment]
Actually now I'm not sure.  It looks like [we have a check for the write barrier](https://github.com/tenderlove/tenderjit/blob/d3f640e6951db41283ce76dbaad25ca7a839daf5/lib/tenderjit/iseq_compiler.rb#L2625).  Maybe the exit isn't working or there's a bug in the predicate method?

**Comment by tenderlove** [comment]
https://github.com/tenderlove/tenderjit/pull/119 might be the fix for this, but I'm not sure.  It was definitely a bug, just not sure if it's *this* bug

**Comment by tenderlove** [comment]
> I enabled VM debugging and this is definitely a GC problem. I saw a crash like this:
> 
> ```
> TenderJIT::PutnilTest#test_putnil = 0.04 s = .                                                                                                                                               
> TenderJIT::InternTest#test_new_symbol = 0.06 s = .                                                                                                                                           
> TenderJIT::InternTest#test_existing_symbol = 0.17 s = .                                                                                                                                      
> TenderJIT::IterFib#test_fib_iter = Assertion Failed: vm.inc:829:vm_exec_core:!RB_TYPE_P(TOPN(0), T_NONE)                                                                                     
> ruby 3.2.0dev (2021-12-29T11:53:43Z master 002fa28599) [x86_64-linux]                                                                                                                        
>                                                                                                                                                                                              
> -- Control frame information -----------------------------------------------                                                                                                                 
> c:0032 p:0038 s:0199 e:000196 BLOCK  /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1743                                                                                           
> c:0031 p:0015 s:0193 e:000192 METHOD /home/aaron/.gem/ruby/3.0.4/bundler/gems/fisk-13a2129aea03/lib/fisk.rb:649                                                                              
> c:0030 p:0065 s:0187 e:000186 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1741                                                                                           
> c:0029 p:0310 s:0173 e:000172 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:130                                                                                            
> c:0028 p:0025 s:0161 e:000160 METHOD /home/aaron/git/tenderjit/lib/tenderjit/iseq_compiler.rb:1023 [FINISH]                                                                                  
> c:0027 p:0000 s:0153 e:000152 METHOD /home/aaron/git/tenderjit/test/tenderjit_test.rb:12                                                                                                     
> c:0026 p:0005 s:0145 e:000144 METHOD /home/aaron/git/tenderjit/test/tend
> ```
> 
> I built Ruby like this:
> 
> ```
> begin
>   set -lx debugflags '-g'
>   set -lx optflags '-O0'
>   set -lx cppflags '-DRUBY_DEBUG -DVM_CHECK_MODE'
>   set -lx CC 'clang'
>   ./configure --prefix=$HOME/.rubies/ruby-trunk --disable-install-doc
> end
> ```
> 
> It enables VM assertions, and one of the assertions is to make sure that we never push T_NONE or T_MOVED on the stack. We're pushing a T_NONE on the stack so clearly something is getting collected that shouldn't be. I'm guessing we're missing a write barrier, as you suspected @64kramsystem.

Also this is probably not true.  I was not feeling well when I wrote this comment, and I had forgotten to update `main` on my Linux machine.  I think the other read barrier fix fixed this particular issue, so sorry about the comment noise 😩

**Comment by 64kramsystem** [comment]
Thanks for the attention :heart_eyes: I'm going to try hard to reproduce it tomorrow :smile:

**Comment by 64kramsystem** [comment]
So! Bad news and good news :grin:

The bad news is that as of current master (b44963e), the error still happens.

Other bad news: it's a sort of Heisenbug. Whenever I add `--debug`, it doesn't reproduce, and when I print statements to stdout, it reproduces much less (:cursing_face:). I think I can investigate more thoroughly though, since using lldb doesn't reduce the reproduction rate.

Better news is that, although it's still nondeterministic, with a certain seed, I can reproduce it around 90% of the times (on my machine and today; I can't speak any other condition, including tomorrow :sweat_smile:).

I attach a couple of execution logs, with the name of the instruction being executed (uglily :grin:) printed out. In both cases, the crash happens on `leave`, after a `putobject`. At a quick glance, the issue happens in the context of a block.

I'll try to investigate further over the next days - xmas holidays make things a bit unpredictable :laughing:

Happy new year! :christmas_tree:

[stack1.txt](https://github.com/tenderlove/tenderjit/files/7796796/stack1.txt)
[stack0.txt](https://github.com/tenderlove/tenderjit/files/7796797/stack0.txt)

**Comment by 64kramsystem** [comment]
St00pid question: In very general terms, is initiating a garbage collection (`GC.start`) supposed to be a safe operation at any time of a TenderJIT given run?

For the sake of ~breaking things~ trying to reproduce the previous error, I've added this:

```diff
diff --git i/lib/tenderjit/iseq_compiler.rb w/lib/tenderjit/iseq_compiler.rb
index 1cc0a46..cad9d65 100644
--- i/lib/tenderjit/iseq_compiler.rb
+++ w/lib/tenderjit/iseq_compiler.rb
@@ -127,6 +127,7 @@ def resume_compiling insn_idx, temp_stack
             print_str("#{sprintf("%04d", @insn_idx)} running   #{name.ljust(LJUST)} #{sprintf("%#x", @iseq.to_i)} SP #{@temp_stack.size}\n")
           end
           @fisk = Fisk.new
+GC.start
           v = send("handle_#{name}", *params)
           if v == :quit
             make_exit(name, @current_pc, @temp_stack).write_to jit_buffer
```

This causes a segfault during the test suite run.

I don't know if the error is expected or not (for example, because in theory it should not happen, but maybe TJ is not handling this currently, etc. ...). If one can't just "put GC.start where they want", sorry for the noise :grin:

If the error shouldn't happen instead, pseudo-randomly polluting the codebase with `GC.start` (or similar) during the build, could be an interesting form of sort-ofuzz testing.

**Comment by tenderlove** [comment]
> In very general terms, is initiating a garbage collection (`GC.start`) supposed to be a safe operation at any time of a TenderJIT given run?

Yes, it should be safe.  This is a good find!  I will give it a try.

**Comment by tenderlove** [comment]
I fixed some problems in #120, but I'm not sure if it's the problem for this ticket.  I'm still seeing a segv with `GC.start` besides the ones that #120 fixed.

**Comment by tenderlove** [comment]
@64kramsystem I added a bunch of fixes in #121.  Hopefully that PR fixes this issue 😅

**Comment by tenderlove** [comment]
I didn't mean to close this

**Comment by 64kramsystem** [comment]
> @64kramsystem I added a bunch of fixes in #121. Hopefully that PR fixes this issue sweat_smile

Amazing work! :star_struck: :star_struck: :star_struck: I can't reproduce the error(s) anymore (both by mass-running the test suite, and by running it with the seed that previously made it fail).

Thanks :pray: 

If you don't experience any error presumingly linked to this, it can be closed :smile:

**Comment by tenderlove** [comment]
I haven't seen any segv errors since #121 so I'll close this.  Thanks so much for helping me find the issues!  I couldn't have done it without the `GC.start` tip 😁
